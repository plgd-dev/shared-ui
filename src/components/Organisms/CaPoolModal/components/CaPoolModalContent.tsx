import React, { FC, useMemo, useState } from 'react'

import Spacer from '../../../Atomic/Spacer'
import Headline from '../../../Atomic/Headline'
import SimpleStripTable from '../../../Atomic/SimpleStripTable'
import { Row as SimpleStripTableRowType } from '../../../Atomic/SimpleStripTable/SimpleStripTable.types'
import { findInEntries, getCommonNameElement, getMenu } from '../utils'
import { Props } from './CaPoolModalContent.types'
import Column from '../../../Atomic/Grid/Column'
import ContentMenu from '../../../Atomic/ContentMenu'
import ContentSwitch from '../../../Atomic/ContentSwitch'
import Row from '../../../Atomic/Grid/Row'

const CaPoolModalContent: FC<Props> = (props) => {
    const { data, i18n, maxHeight } = props

    console.log(data)

    const [activeItem, setActiveItem] = useState('0')

    const menu = useMemo(() => (data ? getMenu(data) : []), [data])
    const hasContentMenu = useMemo(() => data && Array.isArray(data) && data.length > 1, [data])

    return (
        <Row style={{ width: 'calc(100% + 16px)', height: '100%', overflow: 'hidden' }}>
            {hasContentMenu && (
                <Column size={3}>
                    <ContentMenu
                        menuSearch
                        activeItem={activeItem}
                        handleItemClick={(item) => setActiveItem(item.id)}
                        handleSubItemClick={(_item, parentItem) => {
                            _item.contentRef?.current?.scrollIntoView({ behavior: 'smooth' })
                            setActiveItem(parentItem.id)
                        }}
                        menu={menu!}
                        title={i18n.menuTitle}
                    />
                </Column>
            )}
            <Column size={hasContentMenu ? 9 : 12} style={{ height: '100%', overflow: 'auto', maxHeight }}>
                {data && Array.isArray(data) && (
                    <ContentSwitch activeItem={parseInt(activeItem)}>
                        {data?.map((item: any, key: number) => (
                            <div key={key}>
                                <Spacer type='mb-4'>
                                    <Headline type='h6'>{i18n.subjectName}</Headline>
                                </Spacer>
                                <SimpleStripTable leftColSize={6} rightColSize={6} rows={[{ attribute: i18n.commonName, value: item.subject.cn }]} />
                                {item?.issuer?.entries && (
                                    <>
                                        <Spacer type='mt-8 mb-4'>
                                            <Headline type='h6'>{i18n.issuerName}</Headline>
                                        </Spacer>
                                        <SimpleStripTable
                                            leftColSize={6}
                                            rightColSize={6}
                                            rows={[
                                                {
                                                    attribute: i18n.country,
                                                    value: findInEntries(item?.issuer?.entries, 'Country'),
                                                },
                                                {
                                                    attribute: i18n.organization,
                                                    value: findInEntries(item?.issuer?.entries, 'Organization'),
                                                },
                                                {
                                                    attribute: i18n.commonName,
                                                    value: getCommonNameElement(
                                                        findInEntries(item?.issuer?.entries, 'Common Name'),
                                                        activeItem,
                                                        setActiveItem,
                                                        menu
                                                    ),
                                                },
                                            ]}
                                        />
                                    </>
                                )}
                                <Spacer type='mt-8 mb-4'>
                                    <Headline type='h6'>{i18n.validity}</Headline>
                                </Spacer>
                                <SimpleStripTable
                                    leftColSize={6}
                                    rightColSize={6}
                                    rows={[
                                        { attribute: i18n.notBefore, value: item?.notBeforeUTC },
                                        { attribute: i18n.notAfter, value: item?.notAfterUTC },
                                    ]}
                                />
                                {item?.ext?.san?.altNames && (
                                    <>
                                        <Spacer type='mt-8 mb-4'>
                                            <Headline type='h6'>{i18n.subjectAltNames}</Headline>
                                        </Spacer>
                                        <SimpleStripTable
                                            leftColSize={6}
                                            rightColSize={6}
                                            rows={[
                                                {
                                                    attribute: i18n.dNSName,
                                                    value: findInEntries(item?.ext?.san?.altNames, 'DNS Name'),
                                                },
                                            ]}
                                        />
                                    </>
                                )}
                                {item?.subjectPublicKeyInfo && (
                                    <>
                                        <Spacer type='mt-8 mb-4'>
                                            <Headline type='h6'>{i18n.publicKeyInfo}</Headline>
                                        </Spacer>
                                        <SimpleStripTable
                                            leftColSize={6}
                                            rightColSize={6}
                                            rows={
                                                [
                                                    { attribute: i18n.algorithm, value: item?.subjectPublicKeyInfo.kty },
                                                    {
                                                        attribute: i18n.keySize,
                                                        value: item?.subjectPublicKeyInfo.keysize,
                                                    },
                                                    {
                                                        attribute: i18n.exponent,
                                                        value: item?.subjectPublicKeyInfo.keysize,
                                                    },
                                                    item?.subjectPublicKeyInfo.n
                                                        ? {
                                                              attribute: i18n.modules,
                                                              value: (
                                                                  <span
                                                                      style={{
                                                                          overflow: 'hidden',
                                                                          textOverflow: 'ellipsis',
                                                                      }}
                                                                  >
                                                                      {item?.subjectPublicKeyInfo.n}
                                                                  </span>
                                                              ),
                                                          }
                                                        : undefined,
                                                ].filter((i) => !!i) as SimpleStripTableRowType[]
                                            }
                                        />
                                    </>
                                )}
                                <Spacer type='mt-8 mb-4'>
                                    <Headline type='h6'>{i18n.miscellaneous}</Headline>
                                </Spacer>
                                <SimpleStripTable
                                    leftColSize={6}
                                    rightColSize={6}
                                    rows={
                                        [
                                            item?.serialNumber
                                                ? {
                                                      attribute: i18n.serialNumber,
                                                      value: item?.serialNumber,
                                                  }
                                                : undefined,
                                            item?.signature?.name
                                                ? {
                                                      attribute: i18n.signatureAlgorithm,
                                                      value: item?.signature?.name,
                                                  }
                                                : undefined,
                                            { attribute: i18n.version, value: item?.version },
                                            item?.files
                                                ? {
                                                      attribute: i18n.download,
                                                      value: (
                                                          <span>
                                                              <a
                                                                  download='PEM(cert).pem'
                                                                  href={window.URL.createObjectURL(
                                                                      new Blob([item.files.pem.replace(/%0D%0A/g, '\n').replace(/%20/g, ' ')], {
                                                                          type: 'application/x-pem-file',
                                                                      })
                                                                  )}
                                                              >
                                                                  PEM (cert)
                                                              </a>
                                                          </span>
                                                      ),
                                                  }
                                                : undefined,
                                        ].filter((i) => !!i) as SimpleStripTableRowType[]
                                    }
                                />
                                {item?.fingerprint && (item?.fingerprint?.sha256 || item?.fingerprint?.sha1) && (
                                    <>
                                        <Spacer type='mt-8 mb-4'>
                                            <Headline type='h6'>{i18n.fingerprints}</Headline>
                                        </Spacer>
                                        <SimpleStripTable
                                            leftColSize={6}
                                            rightColSize={6}
                                            rows={
                                                [
                                                    item?.fingerprint?.sha256
                                                        ? {
                                                              attribute: 'SHA-256',
                                                              value: (
                                                                  <span
                                                                      style={{
                                                                          overflow: 'hidden',
                                                                          textOverflow: 'ellipsis',
                                                                      }}
                                                                  >
                                                                      {item?.fingerprint?.sha256}
                                                                  </span>
                                                              ),
                                                          }
                                                        : undefined,
                                                    item?.fingerprint?.sha1
                                                        ? {
                                                              attribute: 'SHA-1',
                                                              value: (
                                                                  <span
                                                                      style={{
                                                                          overflow: 'hidden',
                                                                          textOverflow: 'ellipsis',
                                                                      }}
                                                                  >
                                                                      {item?.fingerprint?.sha1}
                                                                  </span>
                                                              ),
                                                          }
                                                        : undefined,
                                                ].filter((i) => !!i) as SimpleStripTableRowType[]
                                            }
                                        />
                                    </>
                                )}
                                {item?.ext?.basicConstraints && (
                                    <>
                                        <Spacer type='mt-8 mb-4'>
                                            <Headline type='h6'>{i18n.basicConstraints}</Headline>
                                        </Spacer>
                                        <SimpleStripTable
                                            leftColSize={6}
                                            rightColSize={6}
                                            rows={[
                                                {
                                                    attribute: i18n.certificateAuthority,
                                                    value: item?.ext?.basicConstraints.cA ? i18n.yes : i18n.no,
                                                },
                                            ]}
                                        />
                                    </>
                                )}
                                {item?.ext?.keyUsages?.purposes && (
                                    <>
                                        <Spacer type='mt-8 mb-4'>
                                            <Headline type='h6'>{i18n.keyUsages}</Headline>
                                        </Spacer>
                                        <SimpleStripTable
                                            leftColSize={6}
                                            rightColSize={6}
                                            rows={[
                                                {
                                                    attribute: i18n.purposes,
                                                    value: item?.ext?.keyUsages.purposes.join(', '),
                                                },
                                            ]}
                                        />
                                    </>
                                )}
                                {item?.ext?.eKeyUsages?.purposes && (
                                    <>
                                        <Spacer type='mt-8 mb-4'>
                                            <Headline type='h6'>{i18n.extendedKeyUsages}</Headline>
                                        </Spacer>
                                        <SimpleStripTable
                                            leftColSize={6}
                                            rightColSize={6}
                                            rows={[
                                                {
                                                    attribute: i18n.purposes,
                                                    value: item?.ext?.eKeyUsages.purposes.join(', '),
                                                },
                                            ]}
                                        />
                                    </>
                                )}
                                {item?.ext?.sKID?.id && (
                                    <>
                                        <Spacer type='mt-8 mb-4'>
                                            <Headline type='h6'>{i18n.subjectKeyID}</Headline>
                                        </Spacer>
                                        <SimpleStripTable
                                            leftColSize={6}
                                            rightColSize={6}
                                            rows={[
                                                {
                                                    attribute: i18n.purposes,
                                                    value: (
                                                        <span
                                                            style={{
                                                                overflow: 'hidden',
                                                                textOverflow: 'ellipsis',
                                                            }}
                                                        >
                                                            {item?.ext?.sKID?.id}
                                                        </span>
                                                    ),
                                                },
                                            ]}
                                        />
                                    </>
                                )}
                                {item?.ext?.aKID?.id && (
                                    <>
                                        <Spacer type='mt-8 mb-4'>
                                            <Headline type='h6'>{i18n.authorityKeyID}</Headline>
                                        </Spacer>
                                        <SimpleStripTable
                                            leftColSize={6}
                                            rightColSize={6}
                                            rows={[
                                                {
                                                    attribute: i18n.purposes,
                                                    value: (
                                                        <span
                                                            style={{
                                                                overflow: 'hidden',
                                                                textOverflow: 'ellipsis',
                                                            }}
                                                        >
                                                            {item?.ext?.aKID?.id}
                                                        </span>
                                                    ),
                                                },
                                            ]}
                                        />
                                    </>
                                )}
                                {item?.ext?.aia?.descriptions && (
                                    <>
                                        <Spacer type='mt-8'>
                                            <Headline type='h6'>{i18n.authorityInfoAIA}</Headline>
                                        </Spacer>
                                        {Object.values(item?.ext?.aia?.descriptions).map((aiaItem: any, key) => (
                                            <Spacer key={key} type={`mt-${key === 0 ? 4 : 6}`}>
                                                <SimpleStripTable
                                                    headerLeft={'#' + `${key + 1}`.padStart(2, '0')}
                                                    leftColSize={6}
                                                    rightColSize={6}
                                                    rows={[
                                                        {
                                                            attribute: i18n.location,
                                                            value: aiaItem.location,
                                                        },
                                                        {
                                                            attribute: i18n.method,
                                                            value: aiaItem.method,
                                                        },
                                                    ]}
                                                />
                                            </Spacer>
                                        ))}
                                    </>
                                )}
                                {item?.ext?.cp?.policies && (
                                    <>
                                        <Spacer type='mt-8'>
                                            <Headline type='h6'>{i18n.certificatePolicies}</Headline>
                                        </Spacer>
                                        {Object.values(item?.ext?.cp?.policies).map((policy: any, key) => (
                                            <Spacer key={key} type={`mt-${key === 0 ? 4 : 6}`}>
                                                <SimpleStripTable
                                                    headerLeft={'#' + `${key + 1}`.padStart(2, '0')}
                                                    leftColSize={6}
                                                    rightColSize={6}
                                                    rows={[
                                                        {
                                                            attribute: i18n.policy,
                                                            value: policy.name,
                                                        },
                                                        {
                                                            attribute: i18n.value,
                                                            value: policy.value,
                                                        },
                                                    ]}
                                                />
                                            </Spacer>
                                        ))}
                                    </>
                                )}
                                {item?.ext?.scts?.timestamps?.length > 0 && (
                                    <>
                                        <Spacer type='mt-8'>
                                            <Headline type='h6'>{i18n.embeddedSCTs}</Headline>
                                        </Spacer>
                                        {item?.ext?.scts?.timestamps.map((timestampItem: any, key: number) => (
                                            <Spacer key={key} type={`mt-${key === 0 ? 4 : 6}`}>
                                                <SimpleStripTable
                                                    headerLeft={'#' + `${key + 1}`.padStart(2, '0')}
                                                    leftColSize={6}
                                                    rightColSize={6}
                                                    rows={[
                                                        {
                                                            attribute: i18n.logID,
                                                            value: (
                                                                <span
                                                                    style={{
                                                                        overflow: 'hidden',
                                                                        textOverflow: 'ellipsis',
                                                                    }}
                                                                >
                                                                    {timestampItem.logId}
                                                                </span>
                                                            ),
                                                        },
                                                        {
                                                            attribute: i18n.signatureAlgorithm,
                                                            value: timestampItem.signatureAlgorithm,
                                                        },
                                                        {
                                                            attribute: i18n.version,
                                                            value: timestampItem.version,
                                                        },
                                                        {
                                                            attribute: i18n.timestamp,
                                                            value: timestampItem.timestampUTC,
                                                        },
                                                    ]}
                                                />
                                            </Spacer>
                                        ))}
                                    </>
                                )}
                            </div>
                        ))}
                    </ContentSwitch>
                )}
            </Column>
        </Row>
    )
}

CaPoolModalContent.displayName = 'CaPoolModalContent'

export default CaPoolModalContent
