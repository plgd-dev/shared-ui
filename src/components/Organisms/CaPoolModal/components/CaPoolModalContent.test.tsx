import { render } from '../../../../../test/jest-wrapper'
import CaPoolModalContent from './CaPoolModalContent'

describe('CaPoolModalContent component', () => {
    const i18n = {
        algorithm: 'Algorithm',
        authorityInfoAIA: 'authorityInfoAIA',
        authorityKeyID: 'authorityKeyID',
        basicConstraints: 'basicConstraints',
        certificateAuthority: 'certificateAuthority',
        certificatePolicies: 'certificatePolicies',
        commonName: 'commonName',
        country: 'country',
        dNSName: 'dNSName',
        download: 'download',
        embeddedSCTs: 'embeddedSCTs',
        exponent: 'exponent',
        extendedKeyUsages: 'extendedKeyUsages',
        fingerprints: 'fingerprints',
        issuerName: 'issuerName',
        keySize: 'keySize',
        keyUsages: 'keyUsages',
        location: 'location',
        logID: 'logID',
        menuTitle: 'menuTitle',
        method: 'method',
        miscellaneous: 'miscellaneous',
        modules: 'modules',
        no: 'no',
        notAfter: 'notAfter',
        notBefore: 'notBefore',
        organization: 'organization',
        policy: 'policy',
        publicKeyInfo: 'publicKeyInfo',
        purposes: 'purposes',
        serialNumber: 'serialNumber',
        signatureAlgorithm: 'signatureAlgorithm',
        subjectAltNames: 'subjectAltNames',
        subjectKeyID: 'subjectKeyID',
        subjectName: 'subjectName',
        timestamp: 'timestamp',
        validity: 'validity',
        value: 'value',
        version: 'version',
        yes: 'yes',
    }

    it('renders without crashing when data is empty', () => {
        const props = {
            data: [],
            i18n,
            maxHeight: 800,
        }
        const { container, asFragment } = render(<CaPoolModalContent {...props} />)
        expect(container).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    // it('renders with data and triggers actions', () => {
    //     const props = {
    //         data: [
    //             {
    //                 ext: {
    //                     aia: {
    //                         critical: false,
    //                     },
    //                     aKID: null,
    //                     basicConstraints: {
    //                         cA: true,
    //                         critical: true,
    //                     },
    //                     cp: {
    //                         critical: false,
    //                     },
    //                     keyUsages: {
    //                         critical: true,
    //                         purposes: ['Certificate Signing', 'CRL Signing'],
    //                     },
    //                     msCrypto: {
    //                         exists: false,
    //                     },
    //                     ocspStaple: {
    //                         critical: false,
    //                         required: false,
    //                     },
    //                     scts: {
    //                         critical: false,
    //                         timestamps: [],
    //                     },
    //                     sKID: {
    //                         critical: false,
    //                         id: 'DE:60:09:A9:5E:1B:00:D5:86:DF:87:62:A9:FC:0A:87:1A:8A:47:F1',
    //                     },
    //                     san: {
    //                         altNames: [],
    //                         critical: false,
    //                     },
    //                 },
    //                 files: {
    //                     pem: '-----BEGIN%20CERTIFICATE-----%0D%0AMIIBZTCCAQqgAwIBAgIQQJcvpwtpUBg42Io1sGFeHDAKBggqhkjOPQQDAjASMRAw%0D%0ADgYDVQQDEwdSb290IENBMB4XDTI0MDEyOTE3Mzg1MFoXDTI1MDEyODE3Mzg1MFow%0D%0AEjEQMA4GA1UEAxMHUm9vdCBDQTBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABA0b%0D%0AeUYp3iFXkU0WU3fFi4knWmU8R1YZXcFded8sufEbx5epZ0g2mdMfi2BnIUMsNUQR%0D%0AP60y/vNMSJfPgoq9Wt+jQjBAMA4GA1UdDwEB/wQEAwIBBjAPBgNVHRMBAf8EBTAD%0D%0AAQH/MB0GA1UdDgQWBBTeYAmpXhsA1Ybfh2Kp/AqHGopH8TAKBggqhkjOPQQDAgNJ%0D%0AADBGAiEAhi4KDlJwkrg2sl72TtG7EwQlGYuw0vn8OjcyLm2kOMUCIQC7TuTHUqbd%0D%0AyKRJYRSVXfYDlrvPyGofp0c+1rRYCJVp2w==%0D%0A-----END%20CERTIFICATE-----%0D%0A',
    //                 },
    //                 fingerprint: {
    //                     sha1: 'CA:1D:41:B7:45:47:5B:C7:E8:9B:F4:D8:26:60:CF:81:BE:49:E1:76',
    //                     sha256: '31:BA:F7:8A:CF:DA:E1:30:A3:12:99:00:99:4C:23:FB:0F:CA:D6:95:6E:56:6D:D9:C7:4A:D6:F6:97:C0:C2:BC',
    //                 },
    //                 issuer: {
    //                     cn: 'Root CA',
    //                     dn: 'cn=Root CA',
    //                     entries: [['Common Name', 'Root CA']],
    //                 },
    //                 notBefore: '29. 1. 2024 18:38:50 (stredoeurópsky letný čas)',
    //                 notBeforeUTC: 'Mon, 29 Jan 2024 17:38:50 GMT',
    //                 notAfter: '28. 1. 2025 18:38:50 (stredoeurópsky letný čas)',
    //                 notAfterUTC: 'Tue, 28 Jan 2025 17:38:50 GMT',
    //                 subject: {
    //                     cn: 'Root CA',
    //                     dn: 'cn=Root CA',
    //                     entries: [['Common Name', 'Root CA']],
    //                 },
    //                 serialNumber: '40:97:2F:A7:0B:69:50:18:38:D8:8A:35:B0:61:5E:1C',
    //                 signature: {
    //                     name: 'ECDSA with SHA-256',
    //                     type: '1.2.840.10045.4.3.2',
    //                 },
    //                 subjectPublicKeyInfo: {
    //                     kty: 'Elliptic Curve',
    //                     keysize: 256,
    //                     x: '0D:1B:79:46:29:DE:21:57:91:4D:16:53:77:C5:8B:89:27:5A:65:3C:47:56:19:5D:C1:5D:79:DF:2C:B9:F1:1B',
    //                     y: 'C7:97:A9:67:48:36:99:D3:1F:8B:60:67:21:43:2C:35:44:11:3F:AD:32:FE:F3:4C:48:97:CF:82:8A:BD:5A:DF',
    //                     xy: '04:0D:1B:79:46:29:DE:21:57:91:4D:16:53:77:C5:8B:89:27:5A:65:3C:47:56:19:5D:C1:5D:79:DF:2C:B9:F1:1B:C7:97:A9:67:48:36:99:D3:1F:8B:60:67:21:43:2C:35:44:11:3F:AD:32:FE:F3:4C:48:97:CF:82:8A:BD:5A:DF',
    //                 },
    //                 unsupportedExtensions: [],
    //                 version: '3',
    //             },
    //         ],
    //         i18n,
    //         maxHeight: '100%',
    //     }
    //
    //     const { getByText } = render(<CaPoolModalContent {...props} />)
    //     expect(getByText('Test CN')).toBeInTheDocument()
    //     expect(getByText('Test Country')).toBeInTheDocument()
    //     expect(getByText('Test Not Before')).toBeInTheDocument()
    //     expect(getByText('Test Not After')).toBeInTheDocument()
    //     expect(getByText('Test DNS Name')).toBeInTheDocument()
    // })
})
