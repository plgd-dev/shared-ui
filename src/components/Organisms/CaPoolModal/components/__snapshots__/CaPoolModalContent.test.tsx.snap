// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaPoolModalContent component renders without crashing when data is empty 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: -16px;
  margin-right: -16px;
  box-sizing: border-box;
}

.emotion-0>* {
  padding-left: 16px;
  padding-right: 16px;
}

.emotion-1 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 100%;
  max-width: 100%;
  position: relative;
  box-sizing: border-box;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 100%;
}

<div
    class="emotion-0"
    style="width: calc(100% + 16px); height: 100%; overflow: hidden;"
  >
    <div
      class="emotion-1"
      style="height: 100%; overflow: auto; max-height: 800px;"
    />
  </div>
</DocumentFragment>
`;
